// @aws-amplify/api-graphql/dist/cjs/internals/generateClient.js
'use strict';

Object.defineProperty(exports, "__esModule", { value: true });
exports.generateClient = void 0;
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
const core_1 = require("@aws-amplify/core");
const v6_1 = require("./v6");
const generateEnumsProperty_1 = require("./utils/generateEnumsProperty");
const generateModelsProperty_1 = require("./utils/generateModelsProperty");
const isApiGraphQLProviderConfig_1 = require("./utils/isApiGraphQLProviderConfig");
const generateCustomOperationsProperty_1 = require("./generateCustomOperationsProperty");
const types_1 = require("../types");
const isConfigureEventWithResourceConfig_1 = require("./utils/isConfigureEventWithResourceConfig");
/**
 * @private
 *
 * Creates a client that can be used to make GraphQL requests, using a provided `AmplifyClassV6`
 * compatible context object for config and auth fetching.
 *
 * @param params
 * @returns
 */
function generateClient(params) {
    const client = {
        [types_1.__amplify]: params.amplify,
        [types_1.__authMode]: params.authMode,
        [types_1.__authToken]: params.authToken,
        [types_1.__headers]: params.headers,
        graphql: v6_1.graphql,
        cancel: v6_1.cancel,
        isCancelError: v6_1.isCancelError,
        models: emptyProperty,
        enums: emptyProperty,
        queries: emptyProperty,
        mutations: emptyProperty,
    };
    const apiGraphqlConfig = params.amplify.getConfig().API?.GraphQL;
    if ((0, isApiGraphQLProviderConfig_1.isApiGraphQLConfig)(apiGraphqlConfig)) {
        client.models = (0, generateModelsProperty_1.generateModelsProperty)(client, apiGraphqlConfig);
        client.enums = (0, generateEnumsProperty_1.generateEnumsProperty)(apiGraphqlConfig);
        client.queries = (0, generateCustomOperationsProperty_1.generateCustomQueriesProperty)(client, apiGraphqlConfig);
        client.mutations = (0, generateCustomOperationsProperty_1.generateCustomMutationsProperty)(client, apiGraphqlConfig);
    }
    else {
        // This happens when the `Amplify.configure()` call gets evaluated after the `generateClient()` call.
        //
        // Cause: when the `generateClient()` and the `Amplify.configure()` calls are located in
        // different source files, script bundlers may randomly arrange their orders in the production
        // bundle.
        //
        // With the current implementation, the `client.models` instance created by `generateClient()`
        // will be rebuilt on every `Amplify.configure()` call that's provided with a valid GraphQL
        // provider configuration.
        //
        // TODO: revisit, and reverify this approach when enabling multiple clients for multi-endpoints
        // configuration.
        generateModelsPropertyOnAmplifyConfigure(client);
    }
    return client;
}
exports.generateClient = generateClient;
const generateModelsPropertyOnAmplifyConfigure = (clientRef) => {
    core_1.Hub.listen('core', coreEvent => {
        if (!(0, isConfigureEventWithResourceConfig_1.isConfigureEventWithResourceConfig)(coreEvent.payload)) {
            return;
        }
        const apiGraphQLConfig = coreEvent.payload.data.API?.GraphQL;
        if ((0, isApiGraphQLProviderConfig_1.isApiGraphQLConfig)(apiGraphQLConfig)) {
            clientRef.models = (0, generateModelsProperty_1.generateModelsProperty)(clientRef, apiGraphQLConfig);
            clientRef.enums = (0, generateEnumsProperty_1.generateEnumsProperty)(apiGraphQLConfig);
            clientRef.queries = (0, generateCustomOperationsProperty_1.generateCustomQueriesProperty)(clientRef, apiGraphQLConfig);
            clientRef.mutations = (0, generateCustomOperationsProperty_1.generateCustomMutationsProperty)(clientRef, apiGraphQLConfig);
        }
    });
};
const emptyProperty = new Proxy({}, {
    get() {
        throw new Error('Client could not be generated. This is likely due to `Amplify.configure()` not being called prior to `generateClient()` or because the configuration passed to `Amplify.configure()` is missing GraphQL provider configuration.');
    },
});
